### PR TITLE
Singularity "Warp" Shader

### PR DESCRIPTION
**Pull request by jamessimo**
### Purpose
This adds a "warp" effect around the singularity using a grabpass shader on the prefab. This is to make the singularity look like its really beanding space time and has the nice effect of objects getting sucked in! 
image

### Notes:
The Singularity has 2 shaders on it. One applies to the "lit" layers and the other only applies on the water or stars layer.
Added a check to mouse utils to make sure the texture can be read. 
Added some other WIP shaders under shaders/experimental. These are not used anywhere but are being worked on. 
Effects may need to be tweaked as we get some user feedback.

### Changelog:
CL: [Improvement] Singularity now looks like its bending space time using shaders!

### CODE REVIEW

**Comment by corp-0** [comment]
does it work nicely over the network? I didn't see any code to network the shader

**Comment by ThatDan123** [comment]
> does it work nicely over the network? I didn't see any code to network the shade

Thats a good point. Singularity stage isnt networked, easy to make it a sync var as its just an enum, and then in the hook make it change the shader stuff.

**Comment by Bod9001** [comment]
my question is, should the shader render over the singularity texture?

**Comment by jamessimo** [comment]
> my question is, should the shader render over the singularity texture?

In can show a version with it over on discord but it make the singularity look deformed and tiny.

**Comment by jamessimo** [comment]
> > does it work nicely over the network? I didn't see any code to network the shade
> 
> Thats a good point. Singularity stage isnt networked, easy to make it a sync var as its just an enum, and then in the hook make it change the shader stuff.

It’s always been like that so not sure if it’s needed since the singularity works currently?

**Comment by jamessimo** [comment]
> does it work nicely over the network? I didn't see any code to network the shader

The shader is just a plane on the singularity, I assumed that as long as it’s inside the prefab it would all still work over network since it’s just inheriting it’s parents pos.

**Review by NoooneyDude — COMMENTED** [review]
Your `UpdateWarpFX` method will only be called from the server (aside from initial spawn via Awake / OnEnable). So the scale won't be synced. You can probably get away with just moving the method call to the `SyncDynamicScale` method.

Did you test with a client / server? Testing in editor means the client _is_ the server, so it'll appear to work.

File: UnityProject/Assets/Scripts/Objects/Singularity.cs [hunk 1]
```diff
@@ -168,6 +168,7 @@ private void SyncDynamicScale(Vector3 oldScale, Vector3 newScale)
 			}
 
 			gameObject.transform.LeanScale(newScale, updateFrequency);
+			UpdateWarpFX(CurrentStage);
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Objects/Singularity.cs]: current stage isnt sync'd

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Objects/Singularity.cs]: You could link the `scaledRadius` and `scaledEffect` by the `Vector3 newScale` was nooneys idea. 

Or could add sync var to stage and then add new hook for it.

File: UnityProject/Assets/Scripts/Objects/Singularity.cs [hunk 2]
```diff
@@ -209,6 +230,8 @@ private void SingularityUpdate()
 			// will sync to clients
 			dynamicScale = GetDynamicScale();
 
+			UpdateWarpFX(CurrentStage);
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Objects/Singularity.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/Objects/Singularity.cs [hunk 3]
```diff
@@ -153,6 +168,12 @@ private void SyncDynamicScale(Vector3 oldScale, Vector3 newScale)
 			}
 
 			gameObject.transform.LeanScale(newScale, updateFrequency);
+			
+		}
+
+		private void SyncCurrentStage(SingularityStages oldStage, SingularityStages newStage)
+		{
+			currentStage = newStage;
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Objects/Singularity.cs:176]:
```suggestion
			currentStage = newStage;
			
			UpdateWarpFX(CurrentStage);
```

**Review by ThatDan123 — APPROVED** [review]

**Review by corp-0 — APPROVED** [review]

**Pull request merged**